packages/runtime/src/next/middleware/request-utils.ts

```diff
@@ -0,0 +1,193 @@
+import { NextRequest, NextResponse } from 'next/server'
+import {
+  InvariantDraftRequestError,
+  MiddlewareError,
+  MissingDraftEndpointError,
+  UnauthorizedDraftRequestError,
+  UnknownDraftFetchRequestError,
+  UnparseableDraftCookieResponseError,
+} from './exceptions'
+
+import { type Cookie, parse as parseSetCookie } from 'set-cookie-parser'
+
+const HeaderNames = {
+  DraftMode: 'X-Makeswift-Draft-Mode',
+  PreviewMode: 'X-Makeswift-Preview-Mode',
+} as const
+
+const SearchParams = {
+  DraftMode: 'x-makeswift-draft-mode',
+  PreviewMode: 'x-makeswift-preview-mode',
+} as const
+
+const CookieRequestEndpoints = {
+  DraftMode: '/api/makeswift/draft-mode',
+  PreviewMode: '/api/makeswift/preview-mode',
+} as const
+
+function getDraftModeSecret(request: NextRequest): string | null {
+  return (
+    request.nextUrl.searchParams.get(SearchParams.DraftMode) ??
+    request.headers.get(HeaderNames.DraftMode) ??
+    null
+  )
+}
+
+function getPreviewModeSecret(request: NextRequest): string | null {
+  return (
+    request.nextUrl.searchParams.get(SearchParams.PreviewMode) ??
+    request.headers.get(HeaderNames.PreviewMode) ??
+    null
+  )
+}
+
+export function isDraftModeRequest(request: NextRequest): boolean {
+  return getDraftModeSecret(request) != null
+}
+
+export function isPreviewModeRequest(request: NextRequest): boolean {
+  return getPreviewModeSecret(request) != null
+}
+
+export function isDraftRequest(request: NextRequest): boolean {
+  return isDraftModeRequest(request) || isPreviewModeRequest(request)
+}
+
+export function getDraftSecret(request: NextRequest): string | null {
+  const draftSecret = getDraftModeSecret(request)
+  const previewSecret = getPreviewModeSecret(request)
+  if (draftSecret != null && previewSecret != null) {
+    throw new InvariantDraftRequestError(
+      "Request can't include draft mode and preview mode secrets",
+    )
+  }
+  return draftSecret ?? previewSecret
+}
+
+async function fetchCookies(url: URL, draftSecret: string): Promise<Cookie[]> {
+  const requestUrl = new URL(url)
+  requestUrl.searchParams.set('secret', encodeURIComponent(draftSecret))
```

Comment:
The `URLSearchParams` API should automatically encode the arguments of `set` so `encodeURIComponent` shouldn't be necessary here